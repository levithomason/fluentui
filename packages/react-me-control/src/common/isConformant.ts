import { isConformant as baseIsConformant, IsConformantOptions } from '@fluentui/react-conformance';

export function isConformant(testInfo: Omit<IsConformantOptions, 'componentPath'>) {
  const defaultOptions = {
    asPropHandlesRef: true,
    componentPath: module!.parent!.filename.replace('.test', ''),
  };

  baseIsConformant(defaultOptions, testInfo);
}
