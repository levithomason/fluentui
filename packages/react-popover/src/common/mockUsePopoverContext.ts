import { PopoverContextValue, usePopoverContext } from '../popoverContext';

/**
 * A test utility to mock the usePopoverContext hook that relies on context selector
 * Don't forget to call `jest.mock(**\/popupContext.ts)` while using this
 * @param options Popover context values to set for testing
 */
export const mockPopoverContext = (options: Partial<PopoverContextValue> = {}) => {
  const mockContext: PopoverContextValue = {
    target: null,
    triggerRef: { current: null },
    contentRef: { current: null },
    open: false,
    setOpen: jest.fn(),
    mountNode: null,
    openOnContext: false,
    openOnHover: false,
    ...options,
  };
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (usePopoverContext as jest.Mock).mockImplementation((selector: (context: PopoverContextValue) => any) => {
    return selector(mockContext);
  });
};
