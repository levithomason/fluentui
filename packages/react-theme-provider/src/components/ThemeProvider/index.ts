export * from './ThemeProvider';
export * from './ThemeProvider.types';
export * from './useThemeProvider';
export * from './renderThemeProvider';
