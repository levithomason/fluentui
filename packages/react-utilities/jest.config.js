let { createConfig } = require('@fluentui/scripts/jest/jest-resources');
let path = require('path');

const config = createConfig({
  moduleNameMapper: require('lerna-alias').jest(),
  setupFiles: [path.resolve(path.join(__dirname, 'config', 'tests.js'))],
});

module.exports = config;
