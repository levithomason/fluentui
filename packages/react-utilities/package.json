{
  "name": "@fluentui/react-utilities",
  "version": "9.0.0-alpha.28",
  "description": "A set of general React-specific utilities.",
  "main": "lib-commonjs/index.js",
  "module": "lib/index.js",
  "typings": "lib/index.d.ts",
  "sideEffects": false,
  "repository": {
    "type": "git",
    "url": "https://github.com/microsoft/fluentui"
  },
  "license": "MIT",
  "scripts": {
    "build": "just-scripts build",
    "clean": "just-scripts clean",
    "code-style": "just-scripts code-style",
    "just": "just-scripts",
    "lint": "just-scripts lint",
    "start-test": "just-scripts jest-watch",
    "test": "just-scripts test",
    "update-snapshots": "just-scripts jest -u"
  },
  "devDependencies": {
    "@fluentui/eslint-plugin": "^1.3.1",
    "@fluentui/scripts": "^1.0.0",
    "@types/enzyme": "3.10.3",
    "@types/enzyme-adapter-react-16": "1.0.3",
    "@types/react": "16.9.42",
    "@types/react-test-renderer": "^16.0.0",
    "enzyme": "~3.10.0",
    "enzyme-adapter-react-16": "^1.15.0",
    "react": "16.8.6",
    "react-test-renderer": "^16.3.0"
  },
  "dependencies": {
    "@fluentui/keyboard-key": "^0.3.2",
    "tslib": "^2.1.0"
  },
  "peerDependencies": {
    "@types/react": ">=16.8.0 <18.0.0",
    "react": ">=16.8.0 <18.0.0"
  },
  "beachball": {
    "tag": "alpha",
    "disallowedChangeTypes": [
      "major",
      "minor",
      "patch"
    ]
  }
}
