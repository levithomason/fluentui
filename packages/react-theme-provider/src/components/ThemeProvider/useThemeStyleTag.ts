import * as React from 'react';
import { useId, usePrevious } from '@fluentui/react-utilities';
import { themeToCSSVariables } from '@fluentui/react-theme';
import { ThemeProviderState } from './ThemeProvider.types';

/**
 * Writes a theme as css variables in a style tag on the provided targetDocument as a rule applied to a CSS class
 *
 * @param theme Theme values that are written as css variables in style rule
 * @returns CSS class to apply the rule
 */
export const useThemeStyleTag = (options: Pick<ThemeProviderState, 'theme' | 'targetDocument'>) => {
  const { targetDocument, theme } = options;
  const styleTagId = useId('theme-provider');
  const styleTag = React.useMemo(() => {
    if (!targetDocument) {
      return null;
    }

    const tag = targetDocument.createElement('style');
    tag.setAttribute('id', styleTagId);
    targetDocument.head.appendChild(tag);
    return tag;
  }, [styleTagId, targetDocument]);

  const cssRule = React.useMemo(() => {
    const cssVars = themeToCSSVariables(theme);
    const cssVarsAsString = Object.keys(cssVars).reduce((cssVarRule, cssVar) => {
      cssVarRule += `${cssVar}: ${cssVars[cssVar]}; `;
      return cssVarRule;
    }, '');

    // result: .theme-provider { --css-var: '#fff' }
    return `.${styleTagId} { ${cssVarsAsString} }`;
  }, [theme, styleTagId]);
  const previousCssRule = usePrevious(cssRule);

  if (styleTag && previousCssRule !== cssRule) {
    const sheet = styleTag.sheet as CSSStyleSheet;

    if (sheet.cssRules.length > 0) {
      sheet.deleteRule(0);
    }

    sheet.insertRule(cssRule, 0);
  }

  // Removes the style tag from the targetDocument on unmount or change
  React.useEffect(() => {
    return () => {
      if (styleTag) {
        // IE11 safe node removal, otherwise use node.remove()
        styleTag.parentElement?.removeChild(styleTag);
      }
    };
  }, [styleTag]);

  return styleTagId;
};
