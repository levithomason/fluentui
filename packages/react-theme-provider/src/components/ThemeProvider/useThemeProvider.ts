import * as React from 'react';
import { makeMergePropsCompat, resolveShorthandProps, useMergedRefs } from '@fluentui/react-utilities';
import { useTheme } from '@fluentui/react-shared-contexts';
import { mergeThemes } from '@fluentui/react-theme';
import { ThemeProviderProps, ThemeProviderState } from './ThemeProvider.types';
import { useThemeStyleTag } from './useThemeStyleTag';

export const themeProviderShorthandProps: (keyof ThemeProviderProps)[] = [];

// eslint-disable-next-line deprecation/deprecation
const mergeProps = makeMergePropsCompat<ThemeProviderState>({ deepMerge: themeProviderShorthandProps });

/**
 * Create the state required to render ThemeProvider.
 *
 * The returned state can be modified with hooks such as useThemeProviderStyles,
 * before being passed to renderThemeProvider.
 *
 * @param props - props from this instance of ThemeProvider
 * @param ref - reference to root HTMLElement of ThemeProvider
 * @param defaultProps - (optional) default prop values provided by the implementing type
 */
export const useThemeProvider = (
  props: ThemeProviderProps,
  ref: React.Ref<HTMLElement>,
  defaultProps?: ThemeProviderProps,
): ThemeProviderState => {
  const state = mergeProps(
    {
      ref: useMergedRefs(ref, React.useRef(null)),
      as: 'div',
      targetDocument: typeof document === 'object' && document,
    },
    defaultProps,
    resolveShorthandProps(props, themeProviderShorthandProps),
  );

  const parentTheme = useTheme();
  const mergedTheme = mergeThemes(parentTheme, state.theme ?? {});

  const themeClassName = useThemeStyleTag({ theme: mergedTheme, targetDocument: state.targetDocument });

  // mergeClasses is not needed here because `themeClassName` is not from a `makeStyles` call
  state.className = [state.className || '', themeClassName].filter(Boolean).join(' ');
  state.theme = mergedTheme;

  return state;
};
