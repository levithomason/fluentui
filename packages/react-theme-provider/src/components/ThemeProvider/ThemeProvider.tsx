import * as React from 'react';
import { ThemeProviderProps } from './ThemeProvider.types';
import { useThemeProvider } from './useThemeProvider';
import { useThemeProviderStyles } from './useThemeProviderStyles';
import { renderThemeProvider } from './renderThemeProvider';

export const ThemeProvider = React.forwardRef<HTMLElement, ThemeProviderProps>((props, ref) => {
  const state = useThemeProvider(props, ref);

  useThemeProviderStyles(state);
  return renderThemeProvider(state);
});

ThemeProvider.displayName = 'ThemeProvider';
