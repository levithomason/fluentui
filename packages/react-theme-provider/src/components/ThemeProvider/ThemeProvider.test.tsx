import * as React from 'react';
import { ThemeContext } from '@fluentui/react-shared-contexts';
import { resetIdsForTests } from '@fluentui/react-utilities';
import * as renderer from 'react-test-renderer';
import { ReactWrapper } from 'enzyme';
import { ThemeProvider } from './ThemeProvider';
import { isConformant } from '../../common/isConformant';

describe('ThemeProvider', () => {
  isConformant({
    disabledTests: ['component-handles-classname'],
    Component: ThemeProvider,
    displayName: 'ThemeProvider',
    helperComponents: [ThemeContext.Provider],
  });

  let wrapper: ReactWrapper | undefined;

  afterEach(() => {
    resetIdsForTests();

    if (wrapper) {
      wrapper.unmount();
      wrapper = undefined;
    }
  });

  /**
   * Note: see more visual regression tests for ThemeProvider in /apps/vr-tests.
   */
  it('renders a default state', () => {
    const component = renderer.create(<ThemeProvider>Default ThemeProvider</ThemeProvider>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
