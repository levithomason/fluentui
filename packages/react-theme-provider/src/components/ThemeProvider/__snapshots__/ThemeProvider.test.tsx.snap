// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeProvider renders a default state 1`] = `
<div
  className="theme-provider1"
>
  Default ThemeProvider
</div>
`;
